plugins/woocommerce/includes/class-wc-cache-helper.php

```diff
@@ -36,6 +39,119 @@ public static function init() {
 		add_action( 'edit_terms', array( __CLASS__, 'clean_term_cache' ), 10, 2 );
 	}
 
+	/**
+	 * Handler for the wp_headers hook, it will try to either produce
+	 * proper cache control headers or return a "304 not modified" response,
+	 * and will fallback to the old cache prevention code if that's not possible.
+	 *
+	 * @param array $headers Original HTTP headers to return.
+	 * @return array HTTP headers to return.
+	 */
+	public static function handle_wp_headers( array $headers ) {
+		return self::handle_page_caching( $headers ) ?? self::prevent_caching( $headers );
+	}
+
+	/**
+	 * Attempt to either return a "304 not modified" response or add proper cache control
+	 * headers to the response if possible.
+	 *
+	 * Currently this method will only act when the requested page is a product page,
+	 * and will act as follows:
+	 *
+	 * - If the request has a "If-Modified-Since" header and it contains a date that is
+	 *   less than the last modification date of the requested product, it will return
+	 *   a "304 not modified" response and finish execution immediately.
+	 * - Otherwise, it will send a "Last-Modified" header appropriate for the product
+	 *   (if that information can be obtained), together with a sensible
+	 *   "Cache-Control" header.
+	 *
+	 * If the request is not for a product page, this method returns null
+	 * so that prevent_caching can take over.
+	 *
+	 * @param array $headers Original HTTP response headers.
+	 * @return array|null New HTTP response headers, or null if no processing is actually performed.
+	 */
+	private static function handle_page_caching( $headers ): ?array {
+		global $wp_query;
+
+		if ( ( ! is_main_query() ) || ( ! is_blog_installed() ) ) {
+			return null;
+		}
+
+		if ( 'product' !== ( $wp_query->query_vars['post_type'] ?? null ) ) {
+			return null;
+		}
+
+		$product_name = $wp_query->query_vars['name'] ?? null;
+		if ( ! $product_name ) {
+			return null;
+		}
+
+		$entity_last_modified_date = wc_get_container()->get( ProductLastModifiedDateCache::class )->get_last_modified_date( $product_name );
+
+		// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$if_modified_since_header_value = wp_unslash( $_SERVER['HTTP_IF_MODIFIED_SINCE'] ?? '' );
+		if ( $if_modified_since_header_value && $entity_last_modified_date ) {
+			$if_modified_since_date = DateTime::createFromFormat( DateTimeInterface::RFC7231, $if_modified_since_header_value );
+			if ( $if_modified_since_date && $if_modified_since_date <= $entity_last_modified_date ) {
+				status_header( 304 );
```

Comment:
I think this is pushing into the territory of caching plugins and we might want to leave this out.